lib/phoenix/endpoint/cowboy2_handler.ex

```diff
@@ -0,0 +1,122 @@
+defmodule Phoenix.Endpoint.Cowboy2Handler do
+  @moduledoc """
+  The Cowboy2 adapter for Phoenix.
+
+  It implements the required `child_spec/3` function as well
+  as the handler for the WebSocket transport.
+
+  ## Custom dispatch options
+
+  You can provide custom dispatch options in order to use Phoenix's
+  builtin Cowboy server with custom handlers. For example, to handle
+  raw WebSockets [as shown in Cowboy's docs](https://github.com/ninenines/cowboy/tree/2.0.x/examples)).
+
+  The options are passed to both `:http` and `:https` keys in the
+  endpoint configuration. However, once you pass your custom dispatch
+  options, you will need to manually wire all Phoenix endpoints,
+  including the socket transports.
+
+  You will need the following rules:
+
+    * Per websocket transport:
+
+          {"/socket/websocket", Phoenix.Endpoint.Cowboy2WebSocket,
+            {Phoenix.Transports.WebSocket,
+              {MyApp.Endpoint, MyApp.UserSocket, :websocket}}}
+
+    * Per longpoll transport:
+
+          {"/socket/long_poll", Plug.Adapters.Cowboy2.Handler,
+            {Phoenix.Transports.LongPoll,
+              {MyApp.Endpoint, MyApp.UserSocket, :longpoll}}}
+
+    * For the endpoint:
+
+          {:_, Plug.Adapters.Cowboy2.Handler, {MyApp.Endpoint, []}}
+
+  For example:
+
+      config :myapp, MyApp.Endpoint,
+        http: [dispatch: [
+                {:_, [
+                    {"/foo", MyApp.CustomHandler, []},
+                    {"/bar", MyApp.AnotherHandler, []},
+                    {:_, Plug.Adapters.Cowboy2.Handler, {MyApp.Endpoint, []}}
+                  ]}]]
+
+  It is also important to specify your handlers first, otherwise
+  Phoenix will intercept the requests before they get to your handler.
+  """
```

Comment:
I made changes to the original in #2643

It fixes formatting (the code blocks don't render completely) and shows how to get the live-reload websocket in there too. 

If that PR gets pulled, can you do the same changes here as well?

Reply:
Yeah, I'll be sure to update it, thanks. Please shout at me if I forget!

Reply:
Just a heads up, it was pulled today.

Reply:
Thanks for the reminder!